File: game/atc-core/server/city.lua
-- ATC City Runtime Bridge
-- Infrastructure, utility grids, traffic signals, environment, resource consumption.
-- City simulation state is server-authoritative.

local API_BASE  = ATC.Config.ApiBase  or 'http://localhost:3000'
local API_TOKEN = ATC.Config.ApiToken or ''

local function apiPost(path, body, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'POST', json.encode(body), {
    ['Content-Type']  = 'application/json',
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

local function apiGet(path, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'GET', '', {
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

local function apiPatch(path, body, cb)
  PerformHttpRequest(API_BASE .. path, function(status, text)
    if cb then cb(status, text and json.decode(text)) end
  end, 'PATCH', json.encode(body), {
    ['Content-Type']  = 'application/json',
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
end

-- ── Public SDK ────────────────────────────────────────────────────────────────

ATC.City = {}

--- Register an infrastructure node (upsert).
--- @param nodeId             string
--- @param nodeName           string
--- @param infrastructureType string  'power_station'|'water_treatment'|'gas_main'|'telecom_hub'|'road_segment'|'bridge'|'tunnel'|'sewage'|'other'
--- @param cb                 function|nil callback(status, node)
function ATC.City.RegisterInfrastructure(nodeId, nodeName, infrastructureType, cb)
  apiPost('/api/v1/city/infrastructure', {
    nodeId             = nodeId,
    nodeName           = nodeName,
    infrastructureType = infrastructureType,
  }, cb)
end

--- Get an infrastructure node by nodeId.
--- @param nodeId string
--- @param cb     function callback(status, node)
function ATC.City.GetInfrastructure(nodeId, cb)
  apiGet('/api/v1/city/infrastructure/' .. nodeId, cb)
end

--- Report an infrastructure failure.
--- @param nodeId       string
--- @param failureType  string  'power_outage'|'water_leak'|'gas_leak'|'road_damage'|'bridge_failure'|'telecom_outage'|'other'
--- @param severity     string  'low'|'medium'|'high'|'critical'
--- @param failureNonce string  idempotency key
--- @param description  string|nil
--- @param cb           function|nil callback(status, failure)
function ATC.City.ReportFailure(nodeId, failureType, severity, failureNonce, description, cb)
  apiPost('/api/v1/city/failures', {
    nodeId       = nodeId,
    failureType  = failureType,
    severity     = severity,
    failureNonce = failureNonce,
    description  = description,
  }, cb)
end

--- Resolve an infrastructure failure.
--- @param failureId  string
--- @param resolvedBy string  principalId
--- @param cb         function|nil callback(status, failure)
function ATC.City.ResolveFailure(failureId, resolvedBy, cb)
  apiPost('/api/v1/city/failures/' .. failureId .. '/resolve', {
    failureId  = failureId,
    resolvedBy = resolvedBy,
  }, cb)
end

--- List all active infrastructure failures.
--- @param cb function callback(status, failures)
function ATC.City.ListActiveFailures(cb)
  apiGet('/api/v1/city/failures/active', cb)
end

--- Update a traffic signal state.
--- @param signalId   string
--- @param signalName string
--- @param state      string  'green'|'yellow'|'red'|'flashing'|'offline'
--- @param changedBy  string|nil  principalId
--- @param cb         function|nil callback(status, signal)
function ATC.City.UpdateTrafficSignal(signalId, signalName, state, changedBy, cb)
  apiPost('/api/v1/city/traffic-signals', {
    signalId   = signalId,
    signalName = signalName,
    state      = state,
    changedBy  = changedBy,
  }, cb)
end

--- Get a traffic signal by signalId.
--- @param signalId string
--- @param cb       function callback(status, signal)
function ATC.City.GetTrafficSignal(signalId, cb)
  apiGet('/api/v1/city/traffic-signals/' .. signalId, cb)
end

--- Update the environment for a region.
--- @param regionId string
--- @param params   table  { weather?, timeOfDay?, temperature?, windSpeed?, visibility?, isEmergencyWeather?, activeEventId? }
--- @param cb       function|nil callback(status, environment)
function ATC.City.UpdateEnvironment(regionId, params, cb)
  local body = params or {}
  body.regionId = regionId
  apiPost('/api/v1/city/environment', body, cb)
end

--- Get the environment state for a region.
--- @param regionId string
--- @param cb       function callback(status, environment)
function ATC.City.GetEnvironment(regionId, cb)
  apiGet('/api/v1/city/environment/' .. regionId, cb)
end

--- Record resource consumption against a utility grid.
--- @param gridId       string
--- @param resourceType string  'power_kwh'|'water_liters'|'gas_m3'|'bandwidth_mb'
--- @param amount       number
--- @param consumerId   string|nil
--- @param periodLabel  string|nil
--- @param cb           function|nil callback(status, record)
function ATC.City.RecordConsumption(gridId, resourceType, amount, consumerId, periodLabel, cb)
  apiPost('/api/v1/city/consumption', {
    gridId       = gridId,
    resourceType = resourceType,
    amount       = amount,
    consumerId   = consumerId,
    periodLabel  = periodLabel,
  }, cb)
end

--- Report a utility grid outage.
--- @param gridId        string
--- @param gridName      string
--- @param utilityType   string  'power'|'water'|'gas'|'telecom'|'sewage'
--- @param outageNonce   string  idempotency key
--- @param reason        string
--- @param affectedZones table|nil  list of zone ids
--- @param cb            function|nil callback(status, grid)
function ATC.City.ReportUtilityOutage(gridId, gridName, utilityType, outageNonce, reason, affectedZones, cb)
  apiPost('/api/v1/city/utility-grids/outage', {
    gridId        = gridId,
    gridName      = gridName,
    utilityType   = utilityType,
    outageNonce   = outageNonce,
    reason        = reason,
    affectedZones = affectedZones or {},
  }, cb)
end

--- Restore a utility grid after an outage.
--- @param gridId                string
--- @param restoredByPrincipalId string
--- @param cb                    function|nil callback(status, grid)
function ATC.City.RestoreUtilityGrid(gridId, restoredByPrincipalId, cb)
  apiPost('/api/v1/city/utility-grids/' .. gridId .. '/restore', {
    restoredByPrincipalId = restoredByPrincipalId,
  }, cb)
end

--- Get a utility grid by gridId.
--- @param gridId string
--- @param cb     function callback(status, grid)
function ATC.City.GetUtilityGrid(gridId, cb)
  apiGet('/api/v1/city/utility-grids/' .. gridId, cb)
end

File: apps/api/src/routes/city.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import { requireCapability } from '../middleware/authorization.js'
import {
  registerInfrastructureSchema,
  updateInfrastructureHealthSchema,
  reportInfrastructureFailureSchema,
  resolveInfrastructureFailureSchema,
  updateTrafficSignalSchema,
  updateEnvironmentSchema,
  recordResourceConsumptionSchema,
  reportUtilityOutageSchema,
  restoreUtilityGridSchema,
} from '@atc/operations'
import { CityRuntimeError } from '@atc/city-runtime'

function cityErrorToStatus(err: CityRuntimeError): number {
  const name = err.constructor.name
  if (
    name === 'UtilityGridAlreadyDownError' ||
    name === 'UtilityGridAlreadyRestoredError' ||
    name === 'InfrastructureAlreadyRecoveredError' ||
    name === 'DuplicateOutageError'
  ) return 409
  if (name.endsWith('NotFoundError')) return 404
  return 500
}

const NOT_CONFIGURED = { error: 'City runtime not configured' }

export async function cityRoutes(
  fastify: FastifyInstance,
  opts: { ctx: AppContext },
): Promise<void> {
  const { ctx } = opts

  // ── Register / upsert infrastructure node ────────────────────────────────────

  fastify.post('/api/v1/city/infrastructure', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.cityInfrastructureService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = registerInfrastructureSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.cityInfrastructureService.updateInfrastructureStatus(
          parsed.data.nodeId,
          parsed.data.nodeName,
          parsed.data.infrastructureType,
          'operational',
          100,
        )
        return reply.status(201).send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Get infrastructure node ───────────────────────────────────────────────────

  fastify.get('/api/v1/city/infrastructure/:nodeId', {
    preHandler: requireCapability(ctx, 'city:read'),
    handler: async (req, reply) => {
      if (!ctx.cityInfrastructureService) return reply.status(503).send(NOT_CONFIGURED)
      const { nodeId } = req.params as { nodeId: string }
      const infra = await ctx.cityInfrastructureService.getInfrastructure(nodeId)
      if (!infra) return reply.status(404).send({ error: 'InfrastructureNotFound' })
      return reply.send(infra)
    },
  })

  // ── Update infrastructure health ──────────────────────────────────────────────

  fastify.patch('/api/v1/city/infrastructure/:nodeId/health', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.cityInfrastructureRepo) return reply.status(503).send(NOT_CONFIGURED)
      const { nodeId } = req.params as { nodeId: string }
      const parsed = updateInfrastructureHealthSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      if (parsed.data.status !== undefined) {
        await ctx.cityInfrastructureRepo.updateStatus(nodeId, parsed.data.status)
      }
      const infra = await ctx.cityInfrastructureRepo.findByNodeId(nodeId)
      if (!infra) return reply.status(404).send({ error: 'InfrastructureNotFound' })
      return reply.send(infra)
    },
  })

  // ── List degraded infrastructure ──────────────────────────────────────────────

  fastify.get('/api/v1/city/infrastructure/degraded', {
    preHandler: requireCapability(ctx, 'city:read'),
    handler: async (req, reply) => {
      if (!ctx.cityInfrastructureService) return reply.status(503).send(NOT_CONFIGURED)
      const degraded = await ctx.cityInfrastructureService.listDegraded()
      return reply.send({ infrastructure: degraded })
    },
  })

  // ── Report infrastructure failure ─────────────────────────────────────────────

  fastify.post('/api/v1/city/failures', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.cityInfrastructureService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = reportInfrastructureFailureSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.cityInfrastructureService.reportFailure({
          nodeId:       parsed.data.nodeId,
          failureType:  parsed.data.failureType,
          severity:     parsed.data.severity,
          failureNonce: parsed.data.failureNonce,
          ...(parsed.data.description !== undefined ? { description: parsed.data.description } : {}),
        })
        return reply.status(201).send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Resolve infrastructure failure ────────────────────────────────────────────

  fastify.post('/api/v1/city/failures/:failureId/resolve', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.infrastructureRecoveryService) return reply.status(503).send(NOT_CONFIGURED)
      const { failureId } = req.params as { failureId: string }
      const parsed = resolveInfrastructureFailureSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.infrastructureRecoveryService.recoverInfrastructure(
          failureId,
          parsed.data.resolvedBy,
        )
        return reply.send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── List active failures ──────────────────────────────────────────────────────

  fastify.get('/api/v1/city/failures/active', {
    preHandler: requireCapability(ctx, 'city:read'),
    handler: async (req, reply) => {
      if (!ctx.infrastructureRecoveryService) return reply.status(503).send(NOT_CONFIGURED)
      const failures = await ctx.infrastructureRecoveryService.listActiveFailures()
      return reply.send({ failures })
    },
  })

  // ── Update traffic signal ─────────────────────────────────────────────────────

  fastify.post('/api/v1/city/traffic-signals', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.trafficSignalService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = updateTrafficSignalSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.trafficSignalService.updateSignal(
          parsed.data.signalId,
          parsed.data.signalName,
          parsed.data.state ?? 'green',
          parsed.data.changedBy,
        )
        return reply.send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Get traffic signal ────────────────────────────────────────────────────────

  fastify.get('/api/v1/city/traffic-signals/:signalId', {
    preHandler: requireCapability(ctx, 'city:read'),
    handler: async (req, reply) => {
      if (!ctx.trafficSignalService) return reply.status(503).send(NOT_CONFIGURED)
      const { signalId } = req.params as { signalId: string }
      const signal = await ctx.trafficSignalService.getSignal(signalId)
      if (!signal) return reply.status(404).send({ error: 'SignalNotFound' })
      return reply.send(signal)
    },
  })

  // ── Update environment ────────────────────────────────────────────────────────

  fastify.post('/api/v1/city/environment', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.environmentRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = updateEnvironmentSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.environmentRuntimeService.updateEnvironment(parsed.data.regionId, {
          ...(parsed.data.weather !== undefined ? { weather: parsed.data.weather } : {}),
          ...(parsed.data.timeOfDay !== undefined ? { timeOfDay: parsed.data.timeOfDay } : {}),
          ...(parsed.data.temperature !== undefined ? { temperature: parsed.data.temperature } : {}),
          ...(parsed.data.windSpeed !== undefined ? { windSpeed: parsed.data.windSpeed } : {}),
          ...(parsed.data.visibility !== undefined ? { visibility: parsed.data.visibility } : {}),
          ...(parsed.data.isEmergencyWeather !== undefined ? { isEmergencyWeather: parsed.data.isEmergencyWeather } : {}),
          ...(parsed.data.activeEventId !== undefined ? { activeEventId: parsed.data.activeEventId } : {}),
        })
        return reply.send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Get environment ───────────────────────────────────────────────────────────

  fastify.get('/api/v1/city/environment/:regionId', {
    preHandler: requireCapability(ctx, 'city:read'),
    handler: async (req, reply) => {
      if (!ctx.environmentRuntimeService) return reply.status(503).send(NOT_CONFIGURED)
      const { regionId } = req.params as { regionId: string }
      const env = await ctx.environmentRuntimeService.getEnvironment(regionId)
      if (!env) return reply.status(404).send({ error: 'EnvironmentNotFound' })
      return reply.send(env)
    },
  })

  // ── Record resource consumption ───────────────────────────────────────────────

  fastify.post('/api/v1/city/consumption', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.resourceConsumptionService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = recordResourceConsumptionSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.resourceConsumptionService.recordConsumption(
          parsed.data.gridId,
          parsed.data.resourceType,
          parsed.data.amount,
          parsed.data.consumerId,
          parsed.data.periodLabel,
        )
        return reply.status(201).send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Report utility outage ─────────────────────────────────────────────────────

  fastify.post('/api/v1/city/utility-grids/outage', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.utilityGridService) return reply.status(503).send(NOT_CONFIGURED)
      const parsed = reportUtilityOutageSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.utilityGridService.reportOutage(
          parsed.data.gridId,
          parsed.data.gridName,
          parsed.data.utilityType,
          parsed.data.outageNonce,
          parsed.data.reason,
          parsed.data.affectedZones,
        )
        return reply.status(201).send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Restore utility grid ──────────────────────────────────────────────────────

  fastify.post('/api/v1/city/utility-grids/:gridId/restore', {
    preHandler: requireCapability(ctx, 'city:write'),
    handler: async (req, reply) => {
      if (!ctx.utilityGridService) return reply.status(503).send(NOT_CONFIGURED)
      const { gridId } = req.params as { gridId: string }
      const parsed = restoreUtilityGridSchema.safeParse(req.body)
      if (!parsed.success) return reply.status(400).send({ error: 'Validation', details: parsed.error.issues })
      try {
        const result = await ctx.utilityGridService.restoreGrid(gridId, parsed.data.restoredByPrincipalId)
        return reply.send(result)
      } catch (err) {
        if (err instanceof CityRuntimeError) return reply.status(cityErrorToStatus(err)).send({ error: err.constructor.name, message: err.message })
        throw err
      }
    },
  })

  // ── Get utility grid ──────────────────────────────────────────────────────────

  fastify.get('/api/v1/city/utility-grids/:gridId', {
    preHandler: requireCapability(ctx, 'city:read'),
    handler: async (req, reply) => {
      if (!ctx.utilityGridService) return reply.status(503).send(NOT_CONFIGURED)
      const { gridId } = req.params as { gridId: string }
      const grid = await ctx.utilityGridService.getGrid(gridId)
      if (!grid) return reply.status(404).send({ error: 'GridNotFound' })
      return reply.send(grid)
    },
  })
}
